zerostash/src/commands/mount.rs

```diff
@@ -256,24 +244,102 @@ impl FilesystemMT for SimpleFs {
         size: u32,
         callback: impl FnOnce(ResultSlice<'_>) -> CallbackResult,
     ) -> CallbackResult {
-        println!("read: {:?} {:#x} @ {:#x}", path, size, offset);
-
-        let file_content = self.file_content.lock().unwrap();
+        debug!("read: {:?} {:#x} @ {:#x}", path, size, offset);
+        let size = size as usize;
+        let offset = offset as usize;
         let real_path = strip_path(path);
         let metadata = self.file_map.get(real_path).unwrap();
-
-        if offset > metadata.size {
+        let metadata_size = metadata.size as usize;
+        let mut chunks = metadata.chunks.clone();
+        chunks.sort_by(|(a, _), (b, _)| a.cmp(b));
+        let mut stack = self.stack.lock().unwrap();
+        let stack = stack
+            .entry(real_path.to_path_buf())
+            .or_insert(RefCell::new((vec![], 0, 0)));
+        let mut stack = stack.borrow_mut();
+        let mut objectreader = self.stash.storage_reader().unwrap();
+
+        if offset > metadata_size {
             return callback(Err(libc::EINVAL));
         }
 
-        let buf: Vec<u8> = match file_content.get(real_path) {
-            Some(buf) => buf.to_vec(),
-            None => return callback(Err(libc::ENOENT)),
-        };
+        let current_read = stack.2;
+        if current_read == offset {
+            let end = usize::min(size, metadata_size - offset);
+            if stack.0.len() < end {
+                loop {
+                    let (start, chunk_p) = chunks.get(stack.1).unwrap();
+                    let arc = (metadata_size as u64, Arc::new(ChunkPointer::default()));
```

Comment:
What happens if you end up reading the null `ChunkPointer`? I *think* this should blow up in an `unwrap()`, but I'm not sure.

Reply:
In what case would a null `ChunkPointer` be returned?

Reply:
Ideally never, but it's a bit of a stretch to depend on implementation instead of the interface. I'd definitely try to encode all behaviour possible instead of relying on "that shouldn't happen".